llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -5538,8 +5539,15 @@ void InstCombinePass::printPipeline(
   OS << '>';
 }
 
+char InstCombinePass::ID = 0;
+
 PreservedAnalyses InstCombinePass::run(Function &F,
                                        FunctionAnalysisManager &AM) {
+  auto &LRT = AM.getResult<LastRunTrackingAnalysis>(F);
+  // No changes since last InstCombine pass, exit early.
+  if (LRT.shouldSkip(&ID, Options))
+    return PreservedAnalyses::all();
```

Comment:
Just a thought, but could we instead bake this into the passmanager itself?

Reply:
I don't agree. Only some canonicalization/cleanup passes benefit from this optimization. Moving the logic into PassManager may introduce higher overhead.

BTW, we cannot do this since InstCombine/SimplifyCFG have pass parameters.

Reply:
agreed about keeping this out of the pass manager, the pass manager shouldn't know which passes are supposed to be idempotent

Reply:
I don't know much about the new pass manager, but conceptually there is a lot of overlap between:
- tracking which analyses are valid, and invalidating them when a pass changes something (and doesn't explicitly preserve that analysis)
- tracking which passes have been run, and no pass since then has changed anything.

In the old pass manager you could _almost_ implement this by making a pass P depend on itself:
```
  AU.addUsedIfAvailableID(ID); // depend on myself
```
and then in its `runOn*` method, check whether it is still "available" from a previous run (obviously you would only do this part in an idempotent pass):
```
  if (mustPreserveAnalysisID(ID)) // is my ID still available?
    return false;
```
Is there any way to do something similar in the new pass manager? I.e. have the pass manager be responsible for tracking which passes are still "available", even if it does not know which ones are idempotent?

Reply:
in the legacy pass manager, analyses were special passes. in the new pass manager, analyses and passes are separate concepts, so that sort of hack doesn't really work. and that doesn't solve the pass parameter issue.

Reply:
Is there no way to get the parameters going into a pass from the new manager?

Reply:
> Is there no way to get the parameters going into a pass from the new manager?

no, I'm not sure how that would work. pass parameters are specified when constructing the pass, none of the pass manager infrastructure can access pass params because that's not really part of the public API of a pass